### ISSUE TITLE
JogAmp Executable slows 3D

### ISSUE THREAD

**Issue by PrinzJuliano** [processing/processing#4783]
I have this issue that jogamp is slowing down the startup of my sketch.
As soon as I close the jogamp_exe_tst......exe (with Taskmanager) it starts in just one or two seconds but leaving it on the start could use two to three minutes.

**Comment by benfry** [comment]
What's jogamp_exe_tst......exe? That's not part of Processing.

**Comment by PrinzJuliano** [comment]
Thats a part of the opengl (Jogl) library that is indeed a part of processing's Java library

**Comment by benfry** [comment]
@codeanticode Have you heard of this? Could this be part of #4239? That JOGL is running some sort of test application? I can't find the .exe anywhere.

**Comment by gohai** [comment]
@xranby You know about this .exe?

**Comment by codeanticode** [comment]
I just run a few OpenGL sketches on a Windows 10 machine, and didn't see any jogamp_exe_tst process. I have also been following #4239, but never able to reproduce it so far.

I found a reference to this exe in this [JOGL issue](https://jogamp.org/bugzilla/show_bug.cgi?id=1219), looks like it is used to call some Windows API methods... Marked as resolved though.

I might be the case that when everything is ok, then it just runs for a brief period of time, but when there is some problem causing to freeze we then see this kind of outcomes, including #4239.

**Comment by PrinzJuliano** [comment]
The API is a Java 3D processing API found on jogamp.org

**Comment by xranby** [comment]
The jogamp_exe_tst is _only_ used by jogamp jogl on windows machines to check if a temporary directory is marked as executable before unpacking the native JNI dll's from the jogamp native jar files. Some anti-virus scanners unfortunally cause issues to run the jogamp_exe_tst.

workaround:
Having all the native JNI dll's stored unpacked. This remove the need for JOGL to first having to use the jogamp_exe_tst to locate an executable temp directory where it can unpack the native JNI dlls.

remove all *-natives-*.jar from the classpath or pass -Djogamp.gluegen.UseTempJarCache=false to disable unpacking of jar files.
then unpack the lib folder from jogamp-all-platforms.7z and rename it natives

A) Traditional Native Library Loading
set -Djava.library.path= to include the directory with the unpacked jogamp JNI dll's from inside natives subfolder for the architecture in use. 

B) Native Library Loading from user.dir/natives/os-arch/ folder
Place the natives folder in the work dir where you launch java from and jogamp will locate the natives there as a fallback without specifying any additional parameters.

if you want to debug jni lib loading then use:

-Djogamp.debug.JNILibLoader=true

**Comment by PrinzJuliano** [comment]
That is good to know. Could you further explain how to apply those measurements? Since I use the processing.exe file for my projects rather than direct Java!

**Comment by benfry** [comment]
Thanks for the background @xranby. Do you think aggressive virus scanners are likely the cause of #4239 in that case?

**Comment by PrinzJuliano** [comment]
Acctually disabling my virus scanner when running any 3D sketch helps drastically! But the problem is the name changes everytime so I can't exclude them from being scanned and withhold. (At least not single file)

**Comment by Sardtok** [comment]
Hmm, I just got a new laptop at work, and I'm having this issue. I'm using the same virus scanner as on my previous laptop, but I never had any issues with Processing on it. The main difference between the old laptop and the current one is the CPU generation and that the new laptop doesn't have a separate GPU, only the integrated one.

**Comment by OctagonHex** [comment]
I have the same problem on my new laptop, a fast Win10 machine, 16 GB Ram, etc.. with integrated graphics card and a mobile NVidia graphics card. I do have some corporate anti-Virus and anti-intrusion Software running, that I can not disable.
I use Processing 3.4. And the same problem occurs with 3.3.7. Last known version without the problem was 2.2.1.
The process jogamp_exe_tst takes around 30 seconds to terminate after it was launched and I think, this is the root-cause for the slow start-up for P3D steches. The jogamp_exe_tst is created in %localappdata%\temp .
@xranby :  Thank you for your suggestions, but I not it to work.
Could you please detail how I can implement the work-around with Processing 3.4 ?
> workaround:
> Having all the native JNI dll's stored unpacked. This remove the need for JOGL to first having to use the jogamp_exe_tst to locate an executable temp directory where it can unpack the native JNI dlls.
> remove all -natives-.jar from the classpath or pass -Djogamp.gluegen.UseTempJarCache=false to disable unpacking of jar files.

First, Processing 3.4 does not come with a .bat file, where I could change these parameters, instead a single .exe is supplied.
Second, I got the command line from the Task-Manager to be like:

'"C:\MyPath\processing-3.4\java\bin\javaw.exe" -Xmx256m -Djava.ext.dirs="C:\MyPath\processing-3.4\\java\\lib\\ext" -Djna.nosys=true
 -Djna.boot.library.path=lib -Djna.nounpack=true -Dsun.java2d.d3d=false -Dsun.java2d.ddoffscreen=false -Dsun.java2d.noddraw=true
 -classpath "lib;lib\pde.jar;core\library\core.jar;lib\jna.jar;lib\jna-platform.jar;lib\antlr.jar;lib\ant.jar;lib\ant-launcher.jar" processing.app.Base'
There are no -natives-.jar in the command-line. where should I remove them instead? 

> then unpack the lib folder from jogamp-all-platforms.7z and rename it natives
> A) Traditional Native Library Loading
> set -Djava.library.path= to include the directory with the unpacked jogamp JNI dll's from inside natives subfolder for the architecture in use.

I extracted the *.dlls from gluegen-rt-natives-windows-amd64.jar and jogl-all-natives-windows-amd64.jar to core/library/native and used
 -Djava.library.path=C:\MyPath\processing-3.4\core\library\native
but that did not change anything.

> B) Native Library Loading from user.dir/natives/os-arch/ folder
> Place the natives folder in the work dir where you launch java from and jogamp will locate the natives there as a fallback without specifying any additional parameters.

What does that translate to for a Windows10 x64 machine?
is the "user.dir" in the user profile directory? %USERPROFILE% or the roaming app data %APPDATA% ?

> if you want to debug jni lib loading then use:
> -Djogamp.debug.JNILibLoader=true
> if you want to debug issues with the windows jogamp_exe_tst then use:
> -Djogamp.debug.IOUtil.Exe

Where (which file or log) do I see the output of this debug settings?

**Comment by xranby** [comment]
after extracting the natives from core\library\gluegen-rt-natives-windows-amd64.jar and core\library\jogl-all-natives-windows-amd64.jar 
you need to delete the two file jogl-all-natives-windows-amd64.jar and gluegen-rt-natives-windows-amd64.jar

**Comment by OctagonHex** [comment]
Thank you for your reply, unfortunatly, this still does not solve my problem.
My current start-up command-line is:

`"C:\MyPath\processing-3.4\java\bin\javaw.exe" -Xmx256m -Djava.ext.dirs="C:\MyPath\processing-3.4\\java\\lib\\ext" -Djna.nosys=true -Djna.boot.library.path=lib -Djna.nounpack=true -Dsun.java2d.d3d=false -Dsun.java2d.ddoffscreen=false -Dsun.java2d.noddraw=true -Djava.library.path=C:\MyPath\processing-3.4\natives\windows-amd64 -Djogamp.gluegen.UseTempJarCache=false -classpath "lib;lib\pde.jar;core\library\core.jar;lib\jna.jar;lib\jna-platform.jar;lib\antlr.jar;lib\ant.jar;lib\ant-launcher.jar" processing.app.Base`

I had to use create the path processing-3.4\natives\windows-amd64 and copy the native *.dlls there, for processing to start my P3D-sketch.
I deleted the files:
```
core\library\gluegen-rt-natives-windows-amd64.jar
core\library\jogl-all-natives-windows-amd64.jar
```

However, the start of my sketch still takes around 45-60 seconds and the jogamp_exe_tst.....exe is still started.
The difference now is, that the temporary directory for unpackung in %localappdata%\Temp\jogamp_0000\file_cache\jln403083125766...\jln515769703297...
now stays empty. This seems encouraging, because my stetch stills runs successfully. 

Do you have an idea, how to disable the execution of jogamp_Exe_tst...exe ?

**Comment by xranby** [comment]
@OctagonHex you are on the right track! In order to disable execution of the jogamp_Exe_tst...exe you need to pass one more command line flag
-Djogamp.gluegen.UseTempJarCache=false

you can add this command line flag into the processing preferences.txt configuration file
locate this file on your system by first open processing: select File -> Preferences 
at the end of the Preferences window there is a link you can click on that will open the filemanager in the foler where this preferences.txt is located, then close processing.

open the preferences.txt and look at the end of the file:
you need to add/change run.options= to
run.options=-Djogamp.gluegen.UseTempJarCache=false

**Comment by OctagonHex** [comment]
Yes! That was the most important step!

Here is my solution for a fresh install for Processing 3.4
Open Processing and click on the file Name for the preferences file. Close Processing.
Open %APPDATA%\Processing\preferences.txt in a text editor.
Search for the line beginning with `run.options=`

**Comment by xranby** [comment]
@OctagonHex do your sketch launch without extra delays now?

**Comment by OctagonHex** [comment]
@xranby : yes, no more delay.   (I edited my post, because I submitted it by accident while typing.)

**Comment by Sardtok** [comment]
@gouessej That would likely require replacing JOGL with another OpenGL Java library like LWJGL. It would require major changes.

**Comment by codeanticode** [comment]
@Startdok we have been working on a LWJGL-library to complement the core, JOGL-based one. It is in its early stages, but is functional:

https://github.com/codeanticode/processing-lwjgl

**Comment by Sardtok** [comment]
That's very cool, @codeanticode! Having LWJGL available could enable making some very fancy libraries, like a Processing OpenCL library, sound playback with OpenAL and GLFW functionality like gamepads.

I'd be interested in rewriting my old LWJGL 2 controller handling library (general event library for HIDs very any input can be mapped to specific actions) to use LWJGL 3's GLFW integration.

**Comment by gouessej** [comment]
@Sardtok You misunderstood my suggestion, maybe it wasn't clear. JOGL uses a tiny binary to get some file permissions, this binary could be replaced by another mechanism calling directly some standard Java APIs available since Java 1.8 which would allow to stick with JOGL and to get rid of the slowness. xranby and me are both JOGL contributors, my suggestion was about improving JOGL, not about replacing it.

**In no way I was suggesting to drop JOGL**. I won't start another "JOGL vs whatever" debate.

**Comment by ralfleggett** [comment]
Sorry to bring up an old issue again - the steps OctagonHex provided work perfectly for running a sketch using the Processing IDE, but unfortunately the same problem occurs when exporting an application (Windows 64bit). Adding the natives directory to the root directory of the application means the application runs after force closing the two jogamp processes with Windows task manager, but I can't find a way to pass the -Djogamp.gluegen.UseTempJarCache=false argument anywhere to stop those processes spawning in the first place. I've tried creating a shortcut that runs the .exe with the option, but that doesn't work. If anyone has a solution that'd be greatly appreciated! Thanks

**Comment by benfry** [comment]
Resolved for 4.0 beta 4, where the libraries will be unpacked to subfolders https://github.com/processing/processing4/commit/be38429ef53b6108dff70855efaaadcab47fdd90
